File: js/src/tokenize.ts
// Lossless tokenizer. `detokenize(tokenize(s)) === s` for every string `s`. Mirrors
// src/pravapis/tokenize.py, reading its word-character tables from generated/chars.json
// instead of a hardcoded copy — see js/src/normalize.ts for the same pattern.
//
// JS Unicode property escapes need `\p{Script=Cyrillic}`, not the bare `\p{Cyrillic}`
// Python's `regex` module accepts; `\p{L}` (General_Category shorthand) is the same in
// both.

import chars from "./generated/chars.json" with { type: "json" };
import type { Token, TokenKind } from "./types.js";

export const APOSTROPHE_CLASS: string = chars.apostrophes.chars.join("");

function escapeForClass(chs: string): string {
  // Only characters that are special inside a [...] class in these strings: - and ^ and ].
  // None of the current apostrophe/letter sets contain them, but a future one might.
  return chs.replace(/[\^\]\\-]/g, "\\$&");
}

const WORD_RE = new RegExp(
  `\\p{Script=Cyrillic}+(?:[${escapeForClass(APOSTROPHE_CLASS)}\\-]\\p{Script=Cyrillic}+)*`,
  "u"
);

const TOKEN_RE = new RegExp(
  `(?<word>\\p{Script=Cyrillic}+(?:[${escapeForClass(APOSTROPHE_CLASS)}\\-]\\p{Script=Cyrillic}+)*)` +
    `|(?<latin>\\p{L}+(?:[${escapeForClass(APOSTROPHE_CLASS)}\\-]\\p{L}+)*)` +
    `|(?<number>\\d+(?:[.,]\\d+)*)` +
    `|(?<space>\\s+)` +
    `|(?<punct>[^\\p{L}\\d\\s]+)` +
    `|(?<other>.)`,
  "gsu"
);

const KIND_BY_GROUP: Record<string, TokenKind> = {
  word: "word",
  latin: "latin",
  number: "number",
  space: "space",
  punct: "punct",
  other: "punct",
};

export function isWordText(text: string): boolean {
  const re = new RegExp(`^${WORD_RE.source}$`, "u");
  return re.test(text);
}

export function tokenize(text: string): Token[] {
  const tokens: Token[] = [];
  TOKEN_RE.lastIndex = 0;
  for (const m of text.matchAll(TOKEN_RE)) {
    const groups = m.groups!;
    const kind = (Object.keys(groups).find((k) => groups[k] !== undefined) ??
      "other") as keyof typeof KIND_BY_GROUP;
    const start = m.index!;
    const end = start + m[0].length;
    tokens.push({ text: m[0], start, end, kind: KIND_BY_GROUP[kind]! });
  }
  return tokens;
}

export function detokenize(tokens: readonly Token[]): string {
  return tokens.map((t) => t.text).join("");
}

// Letters of the Belarusian alphabet, both orthographies (ґ is Taraškievica-only).
export const BELARUSIAN_LETTERS: ReadonlySet<string> = new Set(
  chars.alphabet.pairs.map((p) => p.lower)
);
const WORD_CHARS: ReadonlySet<string> = new Set([...BELARUSIAN_LETTERS, ...APOSTROPHE_CLASS, "-"]);
const UPPER_OF: ReadonlyMap<string, string> = new Map(
  chars.alphabet.pairs.map((p) => [p.lower, p.upper])
);
const NOT_BELARUSIAN: ReadonlySet<string> = new Set([
  ...WORD_CHARS,
  ...[...WORD_CHARS].map((c) => UPPER_OF.get(c) ?? c),
]);

/** True when every letter of `token.text` belongs to the Belarusian alphabet.
 * Russian/Ukrainian words (и, щ, ъ, ї …) return false and are passed through
 * untouched rather than half-converted. */
export function isBelarusianWord(token: Token): boolean {
  if (token.kind !== "word") return false;
  return [...token.text].every((ch) => NOT_BELARUSIAN.has(ch));
}

/** The nearest `window` word tokens on each side of `tokens[i]`. */
export function contextOf(tokens: readonly Token[], i: number, window = 2): Token[] {
  const left: Token[] = [];
  for (let j = i - 1; j >= 0 && left.length < window; j--) {
    if (tokens[j]!.kind === "word") left.push(tokens[j]!);
  }
  left.reverse();
  const right: Token[] = [];
  for (let j = i + 1; j < tokens.length && right.length < window; j++) {
    if (tokens[j]!.kind === "word") right.push(tokens[j]!);
  }
  return [...left, ...right];
}

/** The last word token before index `i` with only whitespace in between. */
export function previousWord(tokens: readonly Token[], i: number): Token | null {
  for (let j = i - 1; j >= 0; j--) {
    if (tokens[j]!.kind === "word") return tokens[j]!;
    if (tokens[j]!.kind !== "space") return null;
  }
  return null;
}

//: Збор 2005 §13 Заўвага: a hyphen (злучок) or a quotation mark (двукоссе) between two
//: words is an invisible bridge to the clitic/conjunction/§18 rules.
export const HYPHENS: ReadonlySet<string> = new Set(["-", "‐", "‑", "­"]);
export const QUOTES: ReadonlySet<string> = new Set([
  "«",
  "»",
  "“",
  "”",
  "„",
  "‟",
  "‹",
  "›",
  '"',
]);

/** Is `tokens[i]` transparent to the §13/§18/clitic cross-word rules? True for
 * whitespace and for quotation marks; true for a hyphen only when it is written
 * against a word (a hyphen with space on both sides is a dash, which blocks the
 * rule). */
export function bridgesWords(tokens: readonly Token[], i: number): boolean {
  const token = tokens[i]!;
  if (token.kind === "space") return true;
  if (token.kind !== "punct") return false;
  const text = token.text;
  if ([...text].every((c) => QUOTES.has(c))) return true;
  if (![...text].every((c) => HYPHENS.has(c) || QUOTES.has(c))) return false;
  const touches =
    (i > 0 && tokens[i - 1]!.kind !== "space") || (i + 1 < tokens.length && tokens[i + 1]!.kind !== "space");
  return touches;
}

/** The first word token after index `i`, across whitespace and §13-transparent
 * punctuation. */
export function nextWord(tokens: readonly Token[], i: number): Token | null {
  for (let j = i + 1; j < tokens.length; j++) {
    if (tokens[j]!.kind === "word") return tokens[j]!;
    if (!bridgesWords(tokens, j)) return null;
  }
  return null;
}

File: src/pravapis/tokenize.py
"""Lossless tokenizer.

``detokenize(tokenize(s)) == s`` for every string ``s`` — the property test
in ``tests/test_properties.py`` enforces it. Every character of the input is
assigned to exactly one token, including whitespace and punctuation, so the
converter can rewrite word tokens in place and reassemble the original layout.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Final

import regex

from pravapis.types import Token, TokenKind

#: Kept in sync with data/chars/apostrophes.tsv (the `char` column) by
#: scripts/check_charclass_sync.py — see that file for why this is a literal here
#: rather than something read off disk.
APOSTROPHE_CLASS: Final[str] = "’'ʼ`‘"

# A Belarusian word: Cyrillic runs joined by apostrophes or hyphens
# (з’ява, па-беларуску, што-небудзь). Trailing/leading hyphens stay punctuation.
WORD_RE: Final[regex.Pattern[str]] = regex.compile(
    rf"\p{{Cyrillic}}+(?:[{APOSTROPHE_CLASS}\-]\p{{Cyrillic}}+)*"
)

_TOKEN_RE: Final[regex.Pattern[str]] = regex.compile(
    rf"(?P<word>\p{{Cyrillic}}+(?:[{APOSTROPHE_CLASS}\-]\p{{Cyrillic}}+)*)"
    rf"|(?P<latin>\p{{L}}+(?:[{APOSTROPHE_CLASS}\-]\p{{L}}+)*)"
    r"|(?P<number>\d+(?:[.,]\d+)*)"
    r"|(?P<space>\s+)"
    r"|(?P<punct>[^\p{L}\d\s]+)"
    r"|(?P<other>.)",
    regex.DOTALL,
)

#: Keyed by ``Match.lastgroup``, which is typed ``str | None``. Every branch of
#: _TOKEN_RE is a named group, so None cannot occur; a KeyError would be loud if it did.
_KIND_BY_GROUP: Final[dict[str | None, TokenKind]] = {
    "word": TokenKind.WORD,
    "latin": TokenKind.LATIN,
    "number": TokenKind.NUMBER,
    "space": TokenKind.SPACE,
    "punct": TokenKind.PUNCT,
    "other": TokenKind.PUNCT,
}

# Letters of the Belarusian alphabet, both orthographies (ґ is Taraškievica-only).
# Kept in sync with data/chars/alphabet.tsv (the `lower` column) by
# scripts/check_charclass_sync.py — that file is NORMATIVE; this is a hand-written
# copy for zero-I/O startup on the hottest path in the pipeline.
BELARUSIAN_LETTERS: Final[frozenset[str]] = frozenset("абвгґдеёжзійклмнопрстуўфхцчшыьэюя")
_WORD_CHARS: Final[frozenset[str]] = BELARUSIAN_LETTERS | frozenset(APOSTROPHE_CLASS + "-")
#: The same set as a negated character class, both cases, so membership is one regex
#: search over the token rather than a genexpr allocating per character.
_NOT_BELARUSIAN: Final[regex.Pattern[str]] = regex.compile(
    "[^" + regex.escape("".join(sorted(_WORD_CHARS | {c.upper() for c in _WORD_CHARS}))) + "]"
)


def tokenize(text: str) -> list[Token]:
    # The hot loop of the whole pipeline: one Token per match over the entire input.
    # Locals for the attribute lookups, span() for one call instead of start()+end(),
    # and no str() around lastgroup, which is already a string.
    tokens: list[Token] = []
    append = tokens.append
    kinds = _KIND_BY_GROUP
    for m in _TOKEN_RE.finditer(text):
        start, end = m.span()
        append(Token(text[start:end], start, end, kinds[m.lastgroup]))
    return tokens


def detokenize(tokens: Sequence[Token]) -> str:
    return "".join(t.text for t in tokens)


def is_belarusian_word(token: Token) -> bool:
    """True when every letter belongs to the Belarusian alphabet.

    Russian/Ukrainian words (и, щ, ъ, ї …) return False and are passed through
    untouched rather than half-converted.
    """
    if token.kind is not TokenKind.WORD:
        return False
    return not _NOT_BELARUSIAN.search(token.text)


def context_of(tokens: Sequence[Token], i: int, window: int = 2) -> list[Token]:
    """The nearest ``window`` word tokens on each side of ``tokens[i]``.

    Non-word tokens are skipped; the token itself is excluded.
    """
    left: list[Token] = []
    j = i - 1
    while j >= 0 and len(left) < window:
        if tokens[j].kind is TokenKind.WORD:
            left.append(tokens[j])
        j -= 1
    left.reverse()
    right: list[Token] = []
    j = i + 1
    while j < len(tokens) and len(right) < window:
        if tokens[j].kind is TokenKind.WORD:
            right.append(tokens[j])
        j += 1
    return left + right


def previous_word(tokens: Sequence[Token], i: int) -> Token | None:
    """The last word token before index ``i`` with only whitespace in between."""
    for j in range(i - 1, -1, -1):
        if tokens[j].kind is TokenKind.WORD:
            return tokens[j]
        if tokens[j].kind is not TokenKind.SPACE:
            return None
    return None


def next_word(tokens: Sequence[Token], i: int) -> Token | None:
    """The first word token after index ``i``, across whitespace and §13 bridges.

    Punctuation breaks the clitic relationship: "не, быў" is not "ня быў". A quotation
    mark is not punctuation for this purpose — §13 Заўвага says so in as many words, and
    the clitic hears what is spoken, where an opening quote is silent. So
    ``з «Віцебскам»`` softens exactly as ``з Віцебскам`` does; be-tarask writes
    зь in both. See :func:`bridges_words` for which marks are transparent.
    """
    for j in range(i + 1, len(tokens)):
        if tokens[j].kind is TokenKind.WORD:
            return tokens[j]
        if not bridges_words(tokens, j):
            return None
    return None


#: Збор 2005 §13 Заўвага: "Злучок і двукоссе не з'яўляюцца знакамі прыпынку і на
#: правапіс й не ўплываюць" — a hyphen (злучок) or a quotation mark (двукоссе) between
#: two words is an invisible bridge, so the conjunction still sees the vowel before it.
#: A dash (працяжнік) is a real punctuation mark and blocks the rule, which is why the
#: en and em dashes are absent here even though they look like long hyphens.
HYPHENS: Final[frozenset[str]] = frozenset("-‐‑­")
QUOTES: Final[frozenset[str]] = frozenset('«»“”„‟‹›"')


def bridges_words(tokens: Sequence[Token], i: int) -> bool:
    """Is ``tokens[i]`` transparent to the §13 conjunction rule?

    True for whitespace and for quotation marks; true for a hyphen only when it is
    written against a word, as in "адна- і шматмоўныя". A hyphen with space on both
    sides is a dash that someone typed on an ASCII keyboard, and a dash blocks the rule.
    """
    token = tokens[i]
    if token.kind is TokenKind.SPACE:
        return True
    if token.kind is not TokenKind.PUNCT:
        return False
    text = token.text
    if all(c in QUOTES for c in text):
        return True
    if not all(c in HYPHENS or c in QUOTES for c in text):
        return False
    touches = (i > 0 and tokens[i - 1].kind is not TokenKind.SPACE) or (
        i + 1 < len(tokens) and tokens[i + 1].kind is not TokenKind.SPACE
    )
    return touches
